pkg/core/klinebuilder.go

```diff
@@ -0,0 +1,221 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/sirupsen/logrus"
+)
+
+// kLineBuilder is a kline builder for a symbol
+// It will accumulate trades and build klines
+//
+//go:generate callbackgen -type kLineBuilder
+type kLineBuilder struct {
+	symbol        string
+	minInterval   types.Interval
+	intervals     map[types.Interval]struct{}
+	isBackTesting bool
+
+	kLineCallbacks       []func(kline types.KLine)
+	kLineClosedCallbacks []func(kline types.KLine)
+	logger               *logrus.Entry
+
+	//
+	mu      sync.Mutex
+	state   *kLineBuilderState
+	running bool
+
+	exchange types.Exchange
+}
+
+// NewKLineBuilder: constructor of kLineBuilder
+//   - symbol: symbol to trace on
+//   - minInterval: unit interval, related to your signal timeframe.
+//     All the supported intervals of the binding stream should be multiple of this interval.
+func NewKLineBuilder(symbol string, minInterval types.Interval, isBackTesting bool) *kLineBuilder {
+	logger := logrus.WithField("symbol", symbol)
+	// Interval.Duration() will panic on invalid value, ex: "-3m"
+	_ = minInterval.Duration()
+
+	return &kLineBuilder{
+		symbol:        symbol,
+		minInterval:   minInterval,
+		isBackTesting: isBackTesting,
+		logger:        logger,
+		intervals:     make(map[types.Interval]struct{}),
+		state:         NewKLineBuilderState(symbol),
+	}
+}
+
+func (kb *kLineBuilder) Subscribe(interval types.Interval) {
+	if interval.Duration()%kb.minInterval.Duration() != 0 {
+		err := fmt.Errorf(
+			"valid interval should be multiple of %s: %s is given",
+			kb.minInterval,
+			interval,
+		)
+		kb.logger.Error(err)
+		panic(err)
+	}
+	kb.intervals[interval] = struct{}{}
+}
+
+func (kb *kLineBuilder) BindStream(stream types.Stream) {
+	if kb.isBackTesting {
+		stream.OnKLineClosed(kb.EmitKLineClosed)
+		stream.OnKLine(kb.EmitKLine)
+		return
+	}
+	stream.OnMarketTrade(kb.handleMarketTrade)
+}
+
+func (kb *kLineBuilder) SetExchange(exchange types.Exchange) {
+	kb.exchange = exchange
+}
+
+func (kb *kLineBuilder) Run(ctx context.Context) {
+	if kb.running {
+		panic("can not run kline builder twice")
+	}
+	if len(kb.intervals) == 0 || kb.isBackTesting {
+		return
+	}
+	kb.logger.Infof("kline updater started for %s (%+v)", kb.symbol, kb.intervals)
+	kb.running = true
+	go kb.run(ctx)
+}
+
+func (kb *kLineBuilder) handleMarketTrade(trade types.Trade) {
+	// the market data stream may subscribe to trades of multiple symbols
+	// so we need to check if the trade is for the symbol we are interested in
+	if trade.Symbol != kb.symbol {
+		return
+	}
+	kb.mu.Lock()
+	defer kb.mu.Unlock()
+
+	kb.state.AddTrade(trade)
+}
+
+func (kb *kLineBuilder) run(ctx context.Context) {
+	startTime := time.Now()
+	startTimeTruncated := startTime.Truncate(kb.minInterval.Duration())
+	intervalDuration := kb.minInterval.Duration()
```

Comment:
since minInterval's purpose is different (for the ticker), we can rename it to tickerInterval, and ensures all the intervals we used are greater than the ticker interval.